var/spack/repos/builtin/packages/hdf5/package.py

```diff
@@ -33,6 +33,10 @@ class Hdf5(CMakePackage):
 
     license("custom")
 
+    language("c")
+    language("cxx", when="+cxx")
+    language("fortran", when="+fortran")
```

Comment:
I like this a lot but I think it would be even clearer (and less cognitive load for users) as:

```python
    depends_on("c")
    depends_on("cxx", when="+cxx")
    depends_on("fortran", when="+fortran")
```

I'm having a hard time understanding why we need `language()` when we could do this.

Reply:
In my opinion we need `language`[^1] because it decouples the information we want to add on language support from the `depends_on` semantic. This has two effects:
1. We can proceed improving runtimes even without having compilers turned into dependencies
2. Later, we have a place where we can extend or change the implementation if need be

To explain point 1, if we used:
```console
depends_on("fortran", when="+fortran")
```
and assumed `fortran` is a virtual, then we would need a provider for that virtual. The provider in the final software model is a compiler, treated as a dependency. This means runtime improvements introducing some information on language support are **blocked** on having compilers as proper dependencies. If we have a directive like `language`, we can adopt an implementation now that fits compilers as node attributes, and change it later to fit compilers as dependencies.[^2]

For point 2, I agree that when we turn compilers into dependencies, we would likely have:
```python
def language(lang_str, when=when_str):
    return depends_on(lang_str, type="build", when=when_str)
```
as a first implementation. The difference with having `depends_on` directly would be that here we have a place where we can decorate or change the implementation if need be. Basically, we differentiate "runtimes" from the start and we have a way to selectively modifying them without modifying packages .

[^1]: By `language` I mean an intermediate directive, I'm not focusing on its name, which can be changed if we have a better one.

[^2]: It's my impression that improvements on runtimes are at the moment much more reachable than compilers as deps, for which we still have some design points to be decided, so this is in my opinion relevant because otherwise we might introduce significant delays.

Reply:
I specifically do not want to differentiate runtimes. They’re dependencies like anything else, and should be modeled that way. This is fundamental to having a general model. I don’t want spack to be like other systems where toolchains are special — that is what we are trying to get away from.

I don’t see why we can’t implement the differentiation *under* depends_on and eventually remove it.

as it is now, a user has to understand why language is different from depends_on. They shouldn’t have to, especially if we do not intend for there to be a difference.

Reply:
I to elaborate on this a little more — I don’t think “languages” are the only things that will need the extra semantics. I think build tools, instrumentation tools, profilers, sanitizers, etc. are going to want to do similar things, and they’re not “languages”. So o think it doesn’t make sense to narrow the meaning of the directive and special case the functionality when it’s really something an arbitrary tool could end up using. It makes more sense to me to keep it general with depends_on.

Reply:
> I specifically do not want to differentiate runtimes.

The runtimes are not differentiated. There's a difference between the language (which brings in the compiler) and the runtimes that are _eventually_ injected by the compiler. 

I can technically do the opposite right now (pseudo-code below):
```python
def depends_on(spec_str, type=("build", "link"), when=when_str):
    if spec_str in SUPPORTED_LANGUAGES:
        # Different treatment of languages until compilers are deps
        assert type == "build"
        return _language(spec_str, type="build", when=when_str)
    
    # Usual code path for other deps
```
but note that this will force us to use the `depends_on` API later. If we need additional arguments, or anything similar, to deal with runtimes at a later stage - and that is not 100% clear to me yet - we'll have to change this API.

Reply:
> I think build tools, instrumentation tools, profilers, sanitizers, etc. are going to want to do similar things, and they’re not “languages”.

Footnote above: "By language I mean an intermediate directive, I'm not focusing on its name, which can be changed if we have a better one."

Reply:
Done in https://github.com/spack/spack/pull/42062/commits/3f930e27a8cae2775feb922b73e414c67ab160de